hypothesis-python/src/hypothesis/internal/constants_ast.py

```diff
@@ -96,22 +96,68 @@ def visit_Constant(self, node):
         self.generic_visit(node)
 
 
-@lru_cache(1024)
-def constants_from_ast(tree: AST) -> AbstractSet[ConstantT]:
-    visitor = ConstantVisitor()
-    visitor.visit(tree)
+def _constants_from_source(source: str) -> AbstractSet[ConstantT]:
+    try:
+        tree = ast.parse(source)
+        visitor = ConstantVisitor()
+        visitor.visit(tree)
+    except Exception:
+        # A bunch of things can go wrong here.
+        # * ast.parse may fail on the source code
+        # * NodeVisitor may hit a RecursionError (see many related issues on
+        #   e.g. libcst https://github.com/Instagram/LibCST/issues?q=recursion),
+        #   or a MemoryError (`"[1, " * 200 + "]" * 200`)
+        return set()
+
     return visitor.constants
 
 
-@lru_cache(1024)
-def _module_ast(module: ModuleType) -> Optional[AST]:
+@lru_cache(4096)
+def constants_from_module(module: ModuleType) -> AbstractSet[ConstantT]:
+    try:
+        module_file = inspect.getsourcefile(module)
+    except Exception:
+        return set()
+
+    if module_file is None:
+        return set()
+
+    try:
+        source_bytes = Path(module_file).read_bytes()
+    except Exception:
+        return set()
+
+    constants_p = (
+        storage_directory("constants") / hashlib.sha1(source_bytes).hexdigest()[:16]
+    )
+    if constants_p.exists():
+        try:
+            source = constants_p.read_text()
+        except Exception:
+            # if there's a problem reading the cached constants, fall back to
+            # standard computation
+            pass
+        else:
+            return _constants_from_source(source)
+
     try:
         source = inspect.getsource(module)
-        tree = ast.parse(source)
     except Exception:
-        return None
+        return set()
+
+    constants = _constants_from_source(source)
+    try:
+        constants_p.parent.mkdir(parents=True, exist_ok=True)
+        constants_p.write_text(
+            f"# file: {module_file}\n# hypothesis_version: {hypothesis.__version__}\n\n"
+            + str(constants),
```

Comment:
mostly-aesthetic thing, but I'd like to sort this set.  Even though the order doesn't matter, it feels pretty weird to vary the cache content by hash-seed when we could make it a proper function of `bytes -> bytes` 😅

Reply:
tiny bit hairy because we need to avoid cross-type comparisons, but I think using a first-index key of `type(v)` buckets by type and avoids this